File: day8/main.py
from __future__ import annotations

from pathlib import Path
import time

import networkx as nx


def load_points(lines: list[str]) -> list[tuple[int, int, int]]:
    pts: list[tuple[int, int, int]] = []
    for line in lines:
        if not line.strip():
            continue
        x, y, z = map(int, line.split(","))
        pts.append((x, y, z))
    return pts


def build_graph(points: list[tuple[int, int, int]]) -> nx.Graph:
    G = nx.Graph()
    n = len(points)
    for i in range(n):
        G.add_node(i)
    for i in range(n):
        xi, yi, zi = points[i]
        for j in range(i + 1, n):
            xj, yj, zj = points[j]
            dx, dy, dz = xi - xj, yi - yj, zi - zj
            dist2 = dx * dx + dy * dy + dz * dz
            G.add_edge(i, j, weight=dist2)
    return G


def part1(G: nx.Graph, k: int = 1000) -> int:
    edges = sorted(G.edges(data=True), key=lambda e: e[2]["weight"])
    H = nx.Graph()
    H.add_nodes_from(G.nodes())
    for u, v, _ in edges[:k]:
        H.add_edge(u, v)
    sizes = sorted([len(c) for c in nx.connected_components(H)], reverse=True)
    while len(sizes) < 3:
        sizes.append(1)
    return sizes[0] * sizes[1] * sizes[2]


def part2(points: list[tuple[int, int, int]], G: nx.Graph) -> int:
    mst = nx.minimum_spanning_tree(G, weight="weight")
    mst_edges = sorted(mst.edges(data=True), key=lambda e: e[2]["weight"])
    if not mst_edges:
        return 0
    a, b, _ = mst_edges[-1]
    return points[a][0] * points[b][0]


def main() -> None:
    lines = Path(__file__).with_name("input.txt").read_text().splitlines()
    points = load_points(lines)
    t0 = time.perf_counter()
    G = build_graph(points)
    p1 = part1(G, k=1000)
    p2 = part2(points, G)
    elapsed_ms = (time.perf_counter() - t0) * 1000
    print(f"top3_product={p1} final_join_x_product={p2} elapsed_ms={elapsed_ms:.3f}")


if __name__ == "__main__":
    main()

File: day8/main.jl
#!/usr/bin/env julia

using Graphs

function load_points(lines)
    pts = []
    for line in lines
        t = strip(line)
        isempty(t) && continue
        x,y,z = parse.(Int, split(t, ","))
        push!(pts, (x,y,z))
    end
    pts
end

function build_graph(pts)
    n = length(pts)
    g = SimpleGraph(n)
    edges_with_weight = []
    for i in 1:n
        xi,yi,zi = pts[i]
        for j in i+1:n
            xj,yj,zj = pts[j]
            dx = xi - xj; dy = yi - yj; dz = zi - zj
            dist2 = dx*dx + dy*dy + dz*dz
            add_edge!(g, i, j)
            push!(edges_with_weight, (dist2, i, j))
        end
    end
    sort!(edges_with_weight, by = x->x[1])
    g, edges_with_weight
end

function part1(n, edges_with_weight; k=1000)
    g = SimpleGraph(n)
    for (idx, (_, a, b)) in enumerate(edges_with_weight)
        idx > k && break
        add_edge!(g, a, b)
    end
    comps = connected_components(g)
    sizes = sort([length(c) for c in comps], rev=true)
    while length(sizes) < 3
        push!(sizes, 1)
    end
    prod(sizes[1:3])
end

function part2(pts, edges_with_weight)
    n = length(pts)
    g = SimpleGraph(n)
    lastprod = 0
    for (_, a, b) in edges_with_weight
        if !has_edge(g, a, b)
            ra = length(connected_components(g))
            add_edge!(g, a, b)
            rb = length(connected_components(g))
            if rb < ra
                lastprod = pts[a][1] * pts[b][1]
                rb == 1 && break
            end
        end
    end
    lastprod
end

function main()
    lines = readlines("input.txt")
    pts = load_points(lines)
    g, edges_with_weight = build_graph(pts)
    t0 = time_ns()
    p1 = part1(length(pts), edges_with_weight, k=1000)
    p2 = part2(pts, edges_with_weight)
    elapsed_ms = (time_ns() - t0)/1e6
    println("top3_product=$(p1) final_join_x_product=$(p2) elapsed_ms=$(round(elapsed_ms; digits=3))")
end

main()
